File: rust_engine/src/fair_value.rs
//! Black-Scholes binary option fair value calculator
//!
//! P(S > K at T) = N(d2)
//! d2 = [ln(S/K) + (r - σ²/2)T] / (σ√T)

/// Standard normal CDF (Abramowitz & Stegun approximation)
pub fn norm_cdf(x: f64) -> f64 {
    let a1 = 0.254829592;
    let a2 = -0.284496736;
    let a3 = 1.421413741;
    let a4 = -1.453152027;
    let a5 = 1.061405429;
    let p = 0.3275911;

    let sign = if x >= 0.0 { 1.0 } else { -1.0 };
    let x = x.abs();
    let t = 1.0 / (1.0 + p * x);
    let y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * (-x * x / 2.0).exp();
    0.5 * (1.0 + sign * y)
}

/// Compute fair value of a binary "BTC > strike" contract
///
/// Returns probability [0.01, 0.99]
pub fn binary_option_price(
    spot: f64,
    strike: f64,
    days_to_expiry: f64,
    volatility: f64,
) -> f64 {
    if spot <= 0.0 || strike <= 0.0 || days_to_expiry <= 0.0 || volatility <= 0.0 {
        return 0.5;
    }

    let t = days_to_expiry / 365.25;
    let r = 0.05; // risk-free rate
    let sigma = volatility;

    let d2 = ((spot / strike).ln() + (r - 0.5 * sigma * sigma) * t) / (sigma * t.sqrt());
    let price = norm_cdf(d2);

    price.clamp(0.01, 0.99)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_atm_option() {
        // ATM: spot = strike, should be ~50%
        let p = binary_option_price(67000.0, 67000.0, 7.0, 0.40);
        assert!(p > 0.45 && p < 0.55, "ATM should be ~50%, got {}", p);
    }

    #[test]
    fn test_deep_itm() {
        // Deep ITM: spot >> strike
        let p = binary_option_price(67000.0, 50000.0, 7.0, 0.40);
        assert!(p > 0.95, "Deep ITM should be >95%, got {}", p);
    }

    #[test]
    fn test_deep_otm() {
        // Deep OTM: spot << strike
        let p = binary_option_price(67000.0, 100000.0, 7.0, 0.40);
        assert!(p < 0.05, "Deep OTM should be <5%, got {}", p);
    }

    #[test]
    fn test_near_expiry() {
        // Near expiry: spot > strike, should be very high
        let p = binary_option_price(67000.0, 60000.0, 0.01, 0.40);
        assert!(p > 0.98, "Near expiry ITM should be >98%, got {}", p);
    }
}

File: src/polymomentum/crypto/fair_value.py
"""Fair value calculator for BTC binary option contracts.

Given: current BTC price, strike price, time to expiry, volatility
Compute: probability that BTC exceeds strike by expiry

Uses Black-Scholes formula for digital (binary) options:
  P(BTC > K at T) = N(d2)
  where d2 = [ln(S/K) + (r - σ²/2)T] / (σ√T)

This is the fair value of a Polymarket "Will BTC exceed $X?" contract.
If the market price diverges from this, there's an arb opportunity.
"""
from __future__ import annotations

import logging
import math
from dataclasses import dataclass

log = logging.getLogger(__name__)


@dataclass
class FairValueResult:
    """Fair value calculation for a binary contract."""
    fair_price: float       # probability [0, 1]
    market_price: float     # current Polymarket price
    edge: float             # fair_price - market_price (positive = buy signal)
    edge_pct: float         # edge as % of market price
    d2: float               # Black-Scholes d2 parameter
    btc_price: float
    strike: float
    days_to_expiry: float
    volatility: float


def _norm_cdf(x: float) -> float:
    """Standard normal CDF using math.erf (ULP-accurate).

    Previously used a hand-rolled Abramowitz & Stegun approximation that had
    systematic errors up to 3.7 percentage points at moderate |x| because the
    polynomial approximates erf(x), not N(x), and the argument must be
    scaled by 1/sqrt(2). We now use Python's stdlib erf which is exact to
    machine precision.

    N(x) = (1 + erf(x / sqrt(2))) / 2
    """
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def compute_fair_value(
    btc_price: float,
    strike: float,
    days_to_expiry: float,
    volatility: float,
    market_price: float = 0.0,
    risk_free_rate: float = 0.05,
) -> FairValueResult:
    """Compute fair value of a binary "BTC > strike" contract.

    Uses Black-Scholes for digital options:
      P(S > K) = N(d2)
      d2 = [ln(S/K) + (r - σ²/2)T] / (σ√T)

    Args:
        btc_price: Current BTC spot price
        strike: Strike price (e.g. 155000)
        days_to_expiry: Days until contract resolution
        volatility: Annualized volatility (e.g. 0.50 = 50%)
        market_price: Current Polymarket price for edge calculation
        risk_free_rate: Annual risk-free rate (default 5%)

    Returns:
        FairValueResult with fair price and edge
    """
    if btc_price <= 0 or strike <= 0 or days_to_expiry <= 0 or volatility < 0.01:
        return FairValueResult(
            fair_price=0.5, market_price=market_price, edge=0,
            edge_pct=0, d2=0, btc_price=btc_price, strike=strike,
            days_to_expiry=days_to_expiry, volatility=volatility,
        )

    T = days_to_expiry / 365.25  # years
    S = btc_price
    K = strike
    r = risk_free_rate
    sigma = volatility

    d2 = (math.log(S / K) + (r - 0.5 * sigma**2) * T) / (sigma * math.sqrt(T))
    fair_price = _norm_cdf(d2)

    # Clamp to [0.01, 0.99] — log when clamping masks extreme values
    raw_fair_price = fair_price
    fair_price = max(0.01, min(0.99, fair_price))
    if raw_fair_price != fair_price:
        log.debug("fair_value.clamped raw=%.4f clamped=%.4f d2=%.2f vol=%.4f T=%.6f",
                  raw_fair_price, fair_price, d2, sigma, T)

    edge = fair_price - market_price
    edge_pct = edge / max(market_price, 0.01)

    return FairValueResult(
        fair_price=fair_price,
        market_price=market_price,
        edge=edge,
        edge_pct=edge_pct,
        d2=d2,
        btc_price=btc_price,
        strike=strike,
        days_to_expiry=days_to_expiry,
        volatility=volatility,
    )
